Mdanalysis

GiovanniBussi opened the pull request:
#### Description

I added the python package MDAnalysis as well as a few packages it depends on.

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.11.6 15G22010
Xcode 8.1 8B62 

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

mf2k commented on python/py-GridDataFormats/Portfile:
```diff
@@ -0,0 +1,41 @@
+                       port:py${python.version}-six \
+                       port:py${python.version}-scipy
+    depends_test-append port:py${python.version}-pytest
+    test.run           yes
```
It's a nitpick for sure. but please:
 - Make sure the above lines conform to the mode line
 - Sort the dependencies alphabetically

mf2k commented on python/py-MDAnalysis/Portfile:
```diff
@@ -0,0 +1,45 @@
+                       port:py${python.version}-scipy \
+                       port:py${python.version}-matplotlib \
+                       port:py${python.version}-mock \
+                       port:py${python.version}-gsd
```
Please sort these dependencies too.

mf2k commented on python/py-mmtf-python/Portfile:
```diff
@@ -0,0 +1,39 @@
+
+if {${name} ne ${subport}} {
+    depends_lib-append port:py${python.version}-msgpack
+    depends_test-append port:py${python.version}-nose
```
Here are some more opportunities to conform to the mode line.

GiovanniBussi commented:
@mf2k thanks for your quick review! I'll certainly fix the dependency order.

Sorry for the dummy question: what does it mean "conforming to the mode line"? Thanks again!

Giovanni

mf2k commented:
I mean that you are supposed to align the right side parameters in an imaginary column every four space. Usually that is 20 spaces but as long as you try to be consistent.

reneeotten commented on python/py-GridDataFormats/Portfile:
```diff
@@ -0,0 +1,41 @@
+                       port:py${python.version}-scipy
+    depends_test-append port:py${python.version}-pytest
+    test.run           yes
+    test.cmd           nosetests-${python.branch}
```
this looks incorrect... you're adding ```pytests``` as a test-dependency, but are calling ```nosetests```?

reneeotten commented on python/py-GridDataFormats/Portfile:
```diff
@@ -0,0 +1,41 @@
+    depends_test-append port:py${python.version}-pytest
+    test.run           yes
+    test.cmd           nosetests-${python.branch}
+    test.target
```
also, do the tests run without having the package already installed? I would imagine that in such case you would actually need to set ```test.env PYTHONPATH=${worksrcpath}/build/lib``` or something similar?

reneeotten commented on python/py-gsd/Portfile:
```diff
@@ -0,0 +1,33 @@
+                    size    198770
+
+if {${name} ne ${subport}} {
+    depends_lib-append port:py${python.version}-numpy
```
you'll need ```py-setuptools``` here as a ```depends_build-append```; haven't checked your the other ones, but I am guessing it's needed there as well. Just look in the ```setup.py``` files to  make sure.

GiovanniBussi commented:
@reneeotten thanks for checking my submission!

I actually found some errors:
- All the new ports were using `setuptools` as you suggested. I added it to the dependencies. Sorry for not noticing it (`pip` does not show it as a dependency).
- One port did not implement any test, I thus removed the test part.
- `py-GridDataFormats` was using `nose`, I fixed it accordingly.

Concerning the `PYTHONPATH` issue, I think this is not a problem. This is what I see testing the `py-GridDataFormats` port:
````
> sudo port deactivate py37-GridDataFormats
Error: port deactivate failed: Image error: port py37-GridDataFormats is not active.
> sudo port clean  py37-GridDataFormats
--->  Cleaning py37-GridDataFormats
> sudo port -d test py37-GridDataFormats
......... etc .........
Ran 29 tests in 1.026s

OK
````

I think this is correct.

@mf2k I think I fixed the spaces so that text start in the 25-th column, it should be correct now.

Thanks!

Giovanni

reneeotten commented on python/py-MDAnalysis/Portfile:
```diff
@@ -0,0 +1,46 @@
+                    size    16457970
+
+if {${name} ne ${subport}} {
+    depends_lib-append  port:py${python.version}-biopython \
```
there is an issue with the use of the compiler here, at least when you follow the description [here](https://trac.macports.org/wiki/UsingTheRightCompiler#testing) it emits the warning that it's using an incorrect compiler.

The issue manifests itself with openmp where the installation scripts tries to determines if OpenMP is installed and whether the compiler supports it (```Attempting to autodetect OpenMP support... Compiler supports OpenMP```). To solve the first issue, one should probably add a variant ```openmp``` and add an explicit dependency on openmp and enable it ; otherwise it will use it in case openmp is installed, but it's not listed as a dependency so it will break when the user uninstalls openmp. The issue with not UsingTheRightCompiler is more complicated and I cannot give any more clues right now, this requires some investigation for which I don't have time right now...

reneeotten replied:
You should be able to reproduce the problem if you have ```libomp`` installed, which you might not have right now...

GiovanniBussi commented:
@reneeotten thanks again for the careful checks!

I think I fixed everything except for the libomp issue.

If I understand your message, the problem is that the script looks for openMP, might find it, but the dependency is not explicitly stated. Actually, on my laptop (OSX 10.11.6), the script uses the system clang compiler (`/usr/bin/clang`) and does not find libomp, even though it is installed.

````
> port installed libomp
The following ports are currently installed:
  libomp @6.0.0_0+universal
  libomp @6.0.1_0+universal
  libomp @7.0.0_0+universal
  libomp @7.0.1_0+universal (active)
````

````
> sudo port install py37-MDAnalysis
....
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_bussi_ports_python_py-MDAnalysis/py37-MDAnalysis/work/MDAnalysis-0.19.2" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build 
Attempting to autodetect OpenMP support... Did not detect OpenMP support.
No openmp compatible compiler found default to serial build.
....
````

I can also confirm that the installed shared object do not link against `libomp`. This command returns an empty string:

````
for file in $(find  /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/MDAnalysis/ -name "*so") ; do otool -L $file ; done | grep libomp
````
(there are no `.dylib` installed, just `.so`).

My experience with MacPorts is that the system compiler never links `libomp`. For a C++ Portfile I am maintaining (`science/plumed/Portfile`) in order to enable openMP I need to specifically use one of the MacPorts compilers (e.g. `+clang60`). The system compiler just does not support the `-fopenmp` flag. So, if my interpretation is correct, `MDAnalysis` should not suffer this problem.

I see that it might make sense to add the 

What do you think?

Giovanni

reneeotten commented:
> @reneeotten thanks again for the careful checks!

 You're welcome! I do note though that I shouldn't need to review trivialities like the license information, that should be straightforward to get right :) I assume that you checked to make sure that my interpretation of the licenses were correct and that this information for the other ports is accurate as well? 

> What do you think?

To me it seems that the Python scripts calls ```gcc``` somewhere irrespective of what MacPorts is telling it do (which should be clang I think). It appears from the log though that it calls ```/usr/bin/clang```.... so I am not sure whether it's a real issue. Anyway, the point is that for me the port actually doesn't install:
```
building 'MDAnalysis.lib.c_distances_openmp' extension
/usr/bin/clang -DNDEBUG -g -fwrapv -O3 -Wall -arch x86_64 -DPARALLEL -I/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/numpy/core/include -IMDAnalysis/lib/include -I/opt/local/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m -c MDAnalysis/lib/c_distances_openmp.c -o build/temp.macosx-10.14-x86_64-3.7/MDAnalysis/lib/c_distances_openmp.o -fopenmp -std=c99 -ffast-math -O3 -funroll-loops
clang: error: unsupported option '-fopenmp'
```

GiovanniBussi commented:
You are totally right, I think I made the mistakes on the Licences since I tried to submit multiple ports at the same time, bad idea...

About the OpenMP issue, I found a way to disable it (latest commit on this branch).

Old Portfile:
````
...
Attempting to autodetect OpenMP support... Did not detect OpenMP support.
No openmp compatible compiler found default to serial build.
Will not attempt to use Cython.
...
````
New Portfile:
````
...
Attempting to autodetect OpenMP support... Did not detect OpenMP support.
Will not attempt to use Cython.
...
````

According to the [setup.py file](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/setup.py), this should disable OpenMP.

When you have time, can you check if this at least compiles on your system?

Thanks!

reneeotten commented on python/py-gsd/Portfile:
```diff
@@ -0,0 +1,34 @@
+                    size    198770
+
+if {${name} ne ${subport}} {
+    depends_build-append port:py${python.version}-setuptools
```
whitespacing is incorrect here; the right column should start at a multiple of 4 space (so not at 26, but either at 25 where you wrap the depends_build-append line or at 29).

reneeotten commented on python/py-GridDataFormats/Portfile:
```diff
@@ -0,0 +1,42 @@
+                    size    63643
+
+if {${name} ne ${subport}} {
+    depends_build-append port:py${python.version}-setuptools
```
whitespacing is incorrect here (as @mfk mentioned before); the right column should start at a multiple of 4 space (so not at 26, but either at 25 where you wrap the depends_build-append line or at 29).

reneeotten commented on python/py-MDAnalysis/Portfile:
```diff
@@ -0,0 +1,47 @@
+                    size    16457970
+
+if {${name} ne ${subport}} {
+    build.env-append     MDA_USE_OPENMP=0
```
whitespacing is incorrect here (as @mfk mentioned before); the right column should start at a multiple of 4 space (so not at 26, but either at 25 where you wrap the depends_build-append line or at 29).

reneeotten commented:
@GiovanniBussi okay I looked a bit more after a cup of coffee and using some other Python ports as an example I think the following will make everything work in ```py-MDAnalysis```:

1) add ```PortGroup           compilers 1.0``` in the beginning of the Portfile
2) define a variant just before the ```if {${name} ne ${subport}} {``` statement:
```
variant openmp description {enable OpenMP parallel acceleration} {
    # Pick a compiler that supports openmp
    compiler.whitelist macports-clang-7.0 macports-clang-6.0 macports-clang-5.0 macports-clang-4.0 \
                        macports-clang-3.9 macports-clang-3.7 macports-gcc-8 macports-gcc-7 macports-gcc-6 \
                        macports-gcc-5 macports-gcc-4.8 macports-gcc-4.7 macports-gcc-4.6 macports-gcc-4.5 \
                        macports-gcc-4.4 macports-gcc-4.3
    compiler.fallback macports-clang-7.0

    build.env-append    CC="${configure.cc}" \
                        CXX="${configure.cxx}" \
                        CFLAGS="${configure.cc_archflags}" \
                        CXXFLAGS="${configure.cxx_archflags}" \
                        CPPFLAGS="${configure.cppflags}"
}
```
3) within the ```if {${name} ne ${subport}} {``` statement add:
```
    if {![ variant_isset openmp ]} {
        build.env-append    MDA_USE_OPENMP=0
    }
```

4) decide whether you want the ```opemp``` as a default variant or not and, if so, add the line
```default_variants +openmp```

In my tests that will work as intended, but please do double-check. I am not completely sure if this is the recommended way of doing it, but there are other ports doing the same and it seems to work ;)

GiovanniBussi commented:
@reneeotten thanks for finding this out!

Meanwhile I pushed a version where a fixed all namespace issues. I was wondering if there is a way to see the incorrect lines while editing the file with VIM. Do you have any idea?

Concerning the compiler thing I tried it on my machine. It compiles correctly using `clang-7.0`, but unfortunately the package cannot find openMP:

````
...
CC='/opt/local/bin/clang-mp-7.0'
...
CXX='/opt/local/bin/clang++-mp-7.0'
...
Executing:  cd "/opt/local/var/macports/build/_Users_bussi_ports_python_py-MDAnalysis/py37-MDAnalysis/work/MDAnalysis-0.19.2" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build 
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_bussi_ports_python_py-MDAnalysis/py37-MDAnalysis/work/MDAnalysis-0.19.2" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build 
Attempting to autodetect OpenMP support... Did not detect OpenMP support.
No openmp compatible compiler found default to serial build.
Will not attempt to use Cython.
...
````

I think there is a problem in the `setup.py` file upstream. I even tried to build the package separately with `python setup.py` and, even though I use an openmp-supporting compiler, openmp is not detected. I am not personally particularly interested in having openmp enabled for this port (I use it for inexpensive analysis or atom selections) so, from my side, it could be merged as is. I added to the `Portfile` a mention to this github issue so that once the problem is solved we can implement your solution.

Let me know if there is anything else I should try.

Thanks again!

Giovanni

cjones051073 commented:
You should not add any variants to enable openmp by hand. So Remove the variant that @reneeotten suggested adding, and instead simply use the mpi portgroup. This will automatically add a number of variants to enable mpi, using a selection of compilers. You can then enable mpi by simply activating one of those variant. If you wish to disable some of them, this can also be done. See for instance

https://github.com/macports/macports-ports/blob/master/python/py-pytorch/Portfile

Where i disable most variants.

cjones051073 commented:
b.t.w. It is expected that using the bare macports clang 7.0 compiler does not enable openmp. To do this you have to use a specific compiler wrapper. This is what the mpi PortGroup does for you. e.g. with pytorch running

```
 sudo port -v build py37-pytorch +clang70 +openmp
```

enables the use of the clang70 based wrapper, that then generates during the build the configuration.

```
-- The CXX compiler identification is Clang 7.0.1
-- The C compiler identification is Clang 7.0.1
-- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70
-- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70
-- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70 -- works
<snip>
-- Found OpenMP_C: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP_CXX: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
<snip>
```

see how the 'compiler' used above is now different...

GiovanniBussi commented:
Hi i am not sure I understand... I do not need openmpi but rather openmp.
Clang-mp-7.0 does support openmp. Am I missing something? Giovanni

Il giorno dom 24 mar 2019 alle 21:34 Chris Jones <notifications@github.com>
ha scritto:

> b.t.w. It is expected that using the bare macports clang 7.0 compiler does
> not enable openmp. To do this you have to use a specific compiler wrapper.
> This is what the mpi PortGroup does for you. e.g. with pytorch running
>
>  sudo port -v build py37-pytorch +clang70 +openmp
>
> enables the use of the clang70 based wrapper, that then generates during
> the build the configuration.
>
> -- The CXX compiler identification is Clang 7.0.1
> -- The C compiler identification is Clang 7.0.1
> -- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70
> -- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70 -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70
> -- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70 -- works
> <snip>
> -- Found OpenMP_C: -fopenmp=libomp (found version "3.1")
> -- Found OpenMP_CXX: -fopenmp=libomp (found version "3.1")
> -- Found OpenMP: TRUE (found version "3.1")
> <snip>
>
> see how the 'compiler' used above is now different...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/macports/macports-ports/pull/3902#issuecomment-475997303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AELRXvFmjiUIN7UklJnxnBOoJcpquQQSks5vZ-FagaJpZM4cB7DW>
> .
>
-- 
Sent from Gmail mobile

cjones051073 commented:
> Hi i am not sure I understand... I do not need openmpi but rather openmp. Clang-mp-7.0 does support openmp. Am I missing something? Giovanni Il giorno dom 24 mar 2019 alle 21:34 Chris Jones <notifications@github.com> ha scritto:

Erm yeah, sorry, guess I read your post too quickly and read it as OpenMPI.

In any case, the mpi PortGroup does also happen to enable openMP at the same time...

Maybe then all you are missing are the `-fopenmp=libmp` options that need to be passed to the compiler ?

> b.t.w. It is expected that using the bare macports clang 7.0 compiler does not enable openmp. To do this you have to use a specific compiler wrapper. This is what the mpi PortGroup does for you. e.g. with pytorch running sudo port -v build py37-pytorch +clang70 +openmp enables the use of the clang70 based wrapper, that then generates during the build the configuration. -- The CXX compiler identification is Clang 7.0.1 -- The C compiler identification is Clang 7.0.1 -- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70 -- Check for working CXX compiler: /opt/local/bin/mpicxx-openmpi-clang70 -- works -- Detecting CXX compiler ABI info -- Detecting CXX compiler ABI info - done -- Detecting CXX compile features -- Detecting CXX compile features - done -- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70 -- Check for working C compiler: /opt/local/bin/mpicc-openmpi-clang70 -- works <snip> -- Found OpenMP_C: -fopenmp=libomp (found version "3.1") -- Found OpenMP_CXX: -fopenmp=libomp (found version "3.1") -- Found OpenMP: TRUE (found version "3.1") <snip> see how the 'compiler' used above is now different... — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#3902 (comment)](https://github.com/macports/macports-ports/pull/3902#issuecomment-475997303)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AELRXvFmjiUIN7UklJnxnBOoJcpquQQSks5vZ-FagaJpZM4cB7DW> .
> -- Sent from Gmail mobile

GiovanniBussi commented:
@cjones051073  I suspect there is a bug in the setup.py file upstream.

Even when I install it using `python setup.py build` openMP is not detected. I will report it to the developers of the package. Meanwhile, I think we can keep the Portfile with openMP disabled.

Thanks!

Giovanni

reneeotten commented:
I am fine with merging this as it is now, with ```openmp``` disabled. If someone feels the need to add this that should be possible after some additional investigation (as I said according to my tests the suggestion I gave worked, but if there is a PortGroup that takes care of this already as mentioned by @cjones051073 that's worth looking into).

cjones051073 commented:
@GiovanniBussi if you are happy to proceed without OpenMP working for now, that is fine by me.
